docs/refs/api/rtos/callbacks.md

```diff
@@ -1,9 +1,384 @@
+public:
+    // Here, the adc_callback_t type is a function that takes in data
+    template <typename T>
+    typedef void (T::*adc_callback_t)(float data);
+
+
+    // In this example, the ADC read function calls the user-provided callback
+    // when data is available.
+    template <typename T>
+    void attach(T *obj, adc_callback_t<T> cb);
+};
+
+class LowPass {
+   float result;
+
+public:
+    // Move the low pass filter implementation to the ADC module
+    void step(float data) {
+        result = result*0.99 + data*0.01;
+    }
+};
+
+
+// Our two adc modules
+ADC adc1;
+ADC adc2;
+
+// Our two low-pass filters
+LowPass low_pass1;
+LowPass low_pass2;
+
+// Here we can register the low-pass filter on both ADC modules
+int main() {
+    // Register one low pass filter
+    adc1.attach(&low_pass1, &LowPass::step);
+    adc2.attach(&low_pass2, &LowPass::step);
+}
+```
+
+#### A separate class for Callbacks)
+
+C++ requires a large set of overloads to support all of the standard function types. It is unreasonable to expect a new library author to add all of these overloads to every function that could take in a callback.
+
+However, C++ provides the tools to delegate this problem to a single class. This class is the Callback class. The Callback class should be familiar to users of the `std::function` class that C++11 introduced but is available for older versions of C++.
+
+**An overly-simplified description of the Callback class is that it contains all of this madness, so you don’t have to.**
+
+Here’s the low-pass filter example rewritten to use the callback class:
+
+``` c++
+class ADC {
+public:
+    // In this example, the ADC read function calls the user-provided callback
+    // when data is available.
+    void attach(Callback<void(float)> cb);
+};
+
+class LowPass {
+   float result;
+
+public:
+    // Move the low pass filter implementation to the ADC module
+    void step(float data) {
+        result = result*0.99 + data*0.01;
+    }
+};
+
+
+// Our two adc modules
+ADC adc1;
+ADC adc2;
+
+// Our two low-pass filters
+LowPass low_pass1;
+LowPass low_pass2;
+
+// Here we can register the low-pass filter on both ADC modules
+int main() {
+    // Register one low pass filter
+    adc1.attach(callback(&low_pass1, &LowPass::step));
+    adc2.attach(callback(&low_pass2, &LowPass::step));
+}
+```
+
+### How to create callbacks
+
+The Callback type is a templated type parameterized by a C++ function declaration:
+
+``` c++
+// Callback</*return type*/(/*parameters*/)> cb;
+Callback<int(float)> cb;          // A callback that takes in a float and returns an int
+Callback<void(float)> cb;         // A callback that takes in a float and returns nothing
+Callback<int()> cb;               // A callback that takes in nothing and returns an int
+Callback<void(float, float)> cb;  // A callback that takes in two floats and returns nothing
+```
+
+You can create a Callback directly from a C function or function pointer with the same type:
```

Comment:
Query: Didn't we already tell them about the content between lines 130 and 154? How is this different from the content above? Can we combine them?

Reply:
Essentially, the following lines are more of an active example whereas the above lines are more of a template.

Reply:
Where did we tell them? The above code blob just covers the type declaration, which is different from the function's type declaration. The two types are independent and if a user is unfamiliar with c function declarations they could be lost.

It's repetitive for the sake of not missing information.